### PR TITLE
drivers: display: Add support for LED matrix driven by nRF SoC GPIOs

### PR DESCRIPTION
**Pull request by anangl**
Add a display driver and the corresponding devicetree binding for a LED
matrix with rows and columns driven by nRF SoCs GPIOs. Such matrix can
be found in the BBC micro:bit v2 board, so a proper devicetree node is
added to its definition so that the driver can be used for that board.

Add also a sample that could be used to see the nRF LED matrix
driver in action and that will prove that the driver is buildable for
both the bbc_microbit_v2 and bbc_microbit boards.

### CODE REVIEW

**Comment by carlescufi** [comment]
@LingaoM @peterniebert @chudov this should be as generic as it gets.

File: boards/arm/bbc_microbit_v2/bbc_microbit_v2.dts [hunk 1]
```diff
@@ -40,6 +40,31 @@
 		};
 	};
 
+	led_matrix {
```
Comment (gmarull) [inline boards/arm/bbc_microbit_v2/bbc_microbit_v2.dts:43]: nit: this addition should be part of a separate commit

File: drivers/display/display_nrf_led_matrix.c [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+		     "Invalid row index in pixel-mapping["#idx"]"); \
+	BUILD_ASSERT((PIXEL_MAPPING(idx) & 0xF) < COL_COUNT, \
+		     "Invalid column index in pixel-mapping["#idx"]");
+DT_FOREACH_PROP_ELEM(MATRIX_NODE, pixel_mapping, CHECK_PIXEL)
+
+#define BASE_FREQUENCY     16000000
+#define TIMER_CLK_CONFIG   NRF_TIMER_FREQ_16MHz
+#define PWM_CLK_CONFIG     NRF_PWM_CLK_16MHz
+#define REFRESH_FREQUENCY  50
+#define BRIGHTNESS_MAX     255
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:43]: some of these parameters, e.g. `REFRESH_FREQUENCY` could be defined in DT.

File: drivers/display/display_nrf_led_matrix.c [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+static void set_pin(uint8_t pin_info, bool active)
+{
+	uint32_t value = active ? 1 : 0;
+
+	if (pin_info & ACTIVE_LOW_MASK) {
+		value = !value;
+	}
+	nrf_gpio_pin_write(pin_info & PSEL_MASK, value);
+}
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:83]: does GPIO API add too much overhead?

File: drivers/display/display_nrf_led_matrix.c [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+
+static int api_blanking_on(const struct device *dev)
+{
+	struct display_drv_data *dev_data = dev->data;
+	const struct display_drv_config *dev_config = dev->config;
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:88]: `data`, `config` are usual var names

File: drivers/display/display_nrf_led_matrix.c [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+	};
+	nrf_pwm_sequence_t sequence = {
+		.values.p_raw = &dev_data->seq,
+		.length = 1,
+	};
+
+	nrf_pwm_pins_set(dev_config->pwm, out_psels);
+	BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:353]: suggestion: I'd centralize build asserts in a single place for better readability

File: drivers/display/display_nrf_led_matrix.c [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+			     NRF_GPIO_PIN_S0S1,
+			     NRF_GPIO_PIN_NOSENSE);
+	}
+
+	for (uint8_t i = 0; i < COL_COUNT; ++i) {
+		uint8_t col_pin_info = dev_config->cols[i];
+
+		set_pin(col_pin_info, false);
+		nrf_gpio_cfg(col_pin_info & PSEL_MASK,
+			     NRF_GPIO_PIN_DIR_OUTPUT,
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:401]: may GPIO API be used here?

File: drivers/display/display_nrf_led_matrix.c [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+static struct display_drv_data instance_data = {
+	.brightness = 0xFF,
+	.blanking   = true,
+};
+
+#define GET_PIN_INFO(node_id, pha, idx) \
+	(DT_GPIO_PIN_BY_IDX(node_id, pha, idx) | \
+	 (DT_PROP_BY_PHANDLE_IDX(node_id, pha, idx, port) << 5) | \
+	 ((DT_GPIO_FLAGS_BY_IDX(node_id, pha, idx) & GPIO_ACTIVE_LOW) ? \
+		ACTIVE_LOW_MASK : 0)),
```
Comment (gmarull) [inline drivers/display/display_nrf_led_matrix.c:433]: ditto for GPIO stuff, gpio_dt_spec and GPIO API may be used

File: dts/bindings/display/nrf-led-matrix.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      description: |
+        Array of GPIOs to be used as rows of the matrix.
+
+    col-gpios:
+      type: phandle-array
+      required: true
+      description: |
+        Array of GPIOs to be used as columns of the matrix.
+
+    pixel-mapping:
```
Comment (gmarull) [inline dts/bindings/display/nrf-led-matrix.yaml:23]: This property could be optional on "regular" LED matrices (like the one in V2). I guess this exists due to the 3x9 matrix used as 5x5 in V1.

File: samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay [hunk 9]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2021, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	led_matrix {
```
Comment (gmarull) [inline samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: can this be added to the board, as for v2?

File: dts/bindings/display/nrf-led-matrix.yaml [hunk 10]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2021, Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Generic LED matrix driven by nRF SoC GPIOs
+
+compatible: "nrf-led-matrix"
```
Comment (gmarull) [inline dts/bindings/display/nrf-led-matrix.yaml:6]: should be `nrf,led-matrix`

File: samples/boards/nrf/nrf_led_matrix/src/main.c [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+		if (ret < 0) {
+			printk("display_write failed: %u/%d\n",
+				__LINE__, ret);
+		}
+
+		ret = display_blanking_off(dev);
+		if (ret < 0) {
+			printk("display_blanking_off failed: %u/%d\n",
+				__LINE__, ret);
+		}
```
Comment (jfischer-no) [inline samples/boards/nrf/nrf_led_matrix/src/main.c:54]: I suggest to move show blocks to own function led_matrix_show_foo(), led_matrix_show_bar()...

File: drivers/display/display_nrf_led_matrix.c [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+		return -ENOTSUP;
+	}
+}
+
+static void api_get_capabilities(const struct device *dev,
+				 struct display_capabilities *caps)
+{
+	caps->x_resolution = X_PIXELS;
+	caps->y_resolution = Y_PIXELS;
+	caps->supported_pixel_formats = PIXEL_FORMAT_MONO10;
```
Comment (jfischer-no) [inline drivers/display/display_nrf_led_matrix.c:185]: PIXEL_FORMAT_MONO01, at least it should match one in api_set_pixel_format(), L:162

File: drivers/display/display_nrf_led_matrix.c [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+#endif
+	.rows = { DT_FOREACH_PROP_ELEM(MATRIX_NODE, row_gpios, GET_PIN_INFO) },
+	.cols = { DT_FOREACH_PROP_ELEM(MATRIX_NODE, col_gpios, GET_PIN_INFO) },
+	.pixel_mapping = DT_PROP(MATRIX_NODE, pixel_mapping),
+};
+
+DEVICE_DT_DEFINE(MATRIX_NODE,
+		 instance_init, NULL,
+		 &instance_data, &instance_config,
+		 APPLICATION, CONFIG_DISPLAY_INIT_PRIORITY, &driver_api);
```
Comment (jfischer-no) [inline drivers/display/display_nrf_led_matrix.c:448]: `POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY, &driver_api);`

**Comment by anangl** [comment]
Rebased.

**Comment by anangl** [comment]
Addressed review comments.

File: drivers/display/display_nrf_led_matrix.c (hunk 2 above)
Comment (anangl) [reply drivers/display/display_nrf_led_matrix.c]: Added the `refresh-frequency` DT property.

File: drivers/display/display_nrf_led_matrix.c (hunk 3 above)
Comment (anangl) [reply drivers/display/display_nrf_led_matrix.c:86]: The overhead is not that important. The actual reason is that I anyway need to operate on PSEL values as I need to provide those in configurations of PWM and GPIOTE channels, see https://github.com/zephyrproject-rtos/zephyr/blob/c3dd5dc6071b6e82d5dbd6395b38e6bfb25911f5/drivers/display/display_nrf_led_matrix.c#L302-L310
And since GPIO interrupts are not used here, I see no reason to make this driver dependent on the GPIO one.

File: drivers/display/display_nrf_led_matrix.c (hunk 7 above)
Comment (anangl) [reply drivers/display/display_nrf_led_matrix.c:435]: See https://github.com/zephyrproject-rtos/zephyr/pull/40968#discussion_r771270367.

File: samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay (hunk 9 above)
Comment (anangl) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: Done.

File: dts/bindings/display/nrf-led-matrix.yaml (hunk 10 above)
Comment (anangl) [reply dts/bindings/display/nrf-led-matrix.yaml]: I used "nordic,nrf-led-matrix". The first part should be the vendor name, right?

File: dts/bindings/display/nrf-led-matrix.yaml (hunk 8 above)
Comment (anangl) [reply dts/bindings/display/nrf-led-matrix.yaml]: Yes, for "regular" matrices, I could generate such array in the driver. But isn't it good to have the mapping specified explicitly in DTS so that everything regarding the hardware is clear?

File: drivers/display/display_nrf_led_matrix.c (hunk 4 above)
Comment (anangl) [reply drivers/display/display_nrf_led_matrix.c:91]: I find them too generic and I had cases in the past where some other `config` or `data` appeared in a given function. So to avoid confusion I'd prefer to keep these prefixes (these are only 4 extra characters). Many nRF drivers actually use such naming.

**Comment by anangl** [comment]
The error and warnings reported by the compliance checks are false positives (in the comparisons, both sides are constant):
```
-:177: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#177: FILE: drivers/display/display_nrf_led_matrix.c:48:
+BUILD_ASSERT(PIXEL_PERIOD <= BIT_MASK(16));

-:179: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#179: FILE: drivers/display/display_nrf_led_matrix.c:50:
+BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);

-:188: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#188: FILE: drivers/display/display_nrf_led_matrix.c:59:
+	NRF_PWM_Type *pwm;
```

File: samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay (hunk 9 above)
Comment (jfischer-no) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: This overlay is no longer needed then.

Comment (anangl) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: Yes, but I kept it is a template, as per:
https://github.com/zephyrproject-rtos/zephyr/blob/c3dd5dc6071b6e82d5dbd6395b38e6bfb25911f5/samples/boards/nrf/nrf_led_matrix/README.rst#L20-L21
That was the reason I haven't originally added the `led_matrix` node to the bbc_microbit board definition. Should I remove the overlay then and that sentence from README?

File: dts/bindings/display/nrf-led-matrix.yaml (hunk 10 above)
Comment (gmarull) [reply dts/bindings/display/nrf-led-matrix.yaml]: correct

File: samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay (hunk 9 above)
Comment (gmarull) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: why not document it on the readme then? having an overlay is a bit confusing if not required.

File: dts/bindings/display/nrf-led-matrix.yaml (hunk 8 above)
Comment (gmarull) [reply dts/bindings/display/nrf-led-matrix.yaml]: fair enough

File: dts/bindings/display/nordic,nrf-led-matrix.yaml [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        - pixel (0,0) is controlled by row 0 and column 0
+        - pixel (1,0) is controlled by row 1 and column 3
+        - pixel (0,1) is controlled by row 2 and column 3
+        - pixel (1,1) is controlled by row 2 and column 4
+        and so on.
+
+    refresh-frequency:
+      type: int
+      required: false
+      default: 50
```
Comment (gmarull) [inline dts/bindings/display/nordic,nrf-led-matrix.yaml:50]: per policy: defaults need to be justified in the description

File: dts/bindings/display/nordic,nrf-led-matrix.yaml [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        specifies that:
+        - pixel (0,0) is controlled by row 0 and column 0
+        - pixel (1,0) is controlled by row 1 and column 3
+        - pixel (0,1) is controlled by row 2 and column 3
+        - pixel (1,1) is controlled by row 2 and column 4
+        and so on.
+
+    refresh-frequency:
+      type: int
+      required: false
```
Comment (gmarull) [inline dts/bindings/display/nordic,nrf-led-matrix.yaml:49]:
```suggestion
```

File: samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay (hunk 9 above)
Comment (gmarull) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: I think the best place for that is the binding as an example (main description field)

Comment (anangl) [reply samples/boards/nrf/nrf_led_matrix/boards/bbc_microbit.overlay:8]: > why not document it on the readme then? having an overlay is a bit confusing if not required.

I agree, it's confusing. I'll remove the overlay.

File: dts/bindings/display/nordic,nrf-led-matrix.yaml (hunk 14 above)
Comment (anangl) [reply dts/bindings/display/nordic,nrf-led-matrix.yaml]: I've made the property required instead.

**Review by gmarull — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Comment by zpm1066** [comment]
@anangl Thank you.
